src/core/packet.lua

```diff
@@ -75,15 +84,49 @@ local function freelist_nfree(freelist)
    return freelist.nfree
 end
 
-local max_packets = 1e6
+local function freelist_lock(freelist)
+   sync.lock(freelist.lock)
+end
+
+local function freelist_unlock(freelist)
+   sync.unlock(freelist.lock)
+end
+
 local packet_allocation_step = 1000
 local packets_allocated = 0
-local packets_fl = ffi.new("struct freelist", max_packets, 0, max_packets)
+local packets_fl = ffi.new("struct freelist")
+      packets_fl.max = max_packets
+local group_fl
+if not shm.exists("group/packets.freelist") then
+   group_fl = shm.create("group/packets.freelist", "struct freelist")
```

Comment:
There is a TOCTTOU here; probably better to always `shm.create`.  `shm.create` does not add `O_EXCL` so it should silently succeed if the file already exists.

Reply:
More broadly, does it always make sense to have a group freelist, even if the system only has one process?  I guess it is OK but it is a point to consider.

Reply:
The code actually works as intended because the group parent will have created `group/packets.freelist` before it starts any children. I.e. `shm.exists("group/packets.freelist")` tests if we are the group parent, which is not something that changes over time.

I agree with the other point about not always having a group freelist, I’ll look into that.